Extend librbdfio.py with a new workloads block. Add minor fixes.

perezjosibm opened the pull request:
This is my first PR, please be gentle in the review. We add the following new features into the CBT for the benchmark librbdfio:

-  Extend the test plan .yml with a new block section 'workloads' to allow a list of performance tests to be indicated, each with its own range of iodepth and numjobs to iterate. An example of this is provided in `example/rbd_fio_test.yml`

- Extend librbdfio.py with support for such 'workloads' block.


### New workloads block
The value of this new workloads block is to allow a finer description of a test job, for example, indicating a range for the iodepth and numjobs FIO parameters, as well as whether to monitor the execution (eg. with perf, top, or collectl). This allow the collection of measurements for computing a response curve.
Here is a snippet that describes:

- a precondition phase consisting on a random write workload, that is not being monitored, then
- a random read workload over the ranges indicated for the FIO parameters numjobs and iodepth
- a random write workload. 

 ```yaml
  workloads:
     precond1:
       jobname: 'precond1rw'
       mode: 'randwrite'
       numjobs: [ 1 ]
       iodepth: [ 8 ]
       monitor: False
     test1:
       jobname: 'rr'
       mode: 'randread'
       numjobs: [ 1, 4, 8 ]
       iodepth: [ 1, 4, 16]
     test2:
       jobname: 'rw'
       mode: 'randwrite'
       numjobs: [ 4 ]
       iodepth: [ 16 ]
```

The following snippet shows the execution of the precondition phase:
![Screenshot 2024-05-07 at 16 07 38](https://github.com/ceph/cbt/assets/23522684/b7524269-8ebd-4628-bb42-2ab0a7672b15)

The following snippet shows the execution of the tests1 and 2:
![Screenshot 2024-05-07 at 16 08 14](https://github.com/ceph/cbt/assets/23522684/32f8060e-f73b-4fd1-80c2-b5bef19da19a)

The following snippet shows the resulting files from the execution of the test:
![Screenshot 2024-05-13 at 11 56 20](https://github.com/ceph/cbt/assets/23522684/93c423d7-957a-48e4-bee8-17d5a578a824)

perezjosibm commented:
@markhpc : here is my PR for the CBT extensions, would you be able to review please? thanks!

sseshasa commented:
@perezjosibm I have started taking a look at this. Please allow me some time to understand the changes. I plan to complete the review by the end of this week or in the worst case early next week due to the long weekend here.

sseshasa reviewed (changes requested):
I have added my first set of review comments. Will continue the review.

sseshasa commented:
> I have added my first set of review comments. Will continue the review.

Hmm...looks like my initial comments were lost. I will add them again. Sorry about that.

sseshasa commented on benchmark/benchmark.py:
```diff
@@ -22,6 +22,7 @@ def __init__(self, archive_dir, cluster, config):
                                         'results',
                                         '{:0>8}'.format(config.get('iteration')),
                                         'id-{}'.format(digest))
+        logger.info("Results dir: %s", self.archive_dir ),
```
Is this log necessary since the archive location is known to the user?

perezjosibm replied:
Being a user for the first time to CBT I was not aware of the location, I can replace it with a debug log sentence instead.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -31,25 +32,34 @@ def __init__(self, archive_dir, cluster, config):
         self.rwmixwrite = 100 - self.rwmixread
         self.log_avg_msec = config.get('log_avg_msec', None)
-#        self.ioengine = config.get('ioengine', 'libaio')
+        self.ioengine = config.get('ioengine', 'libaio')
```
For librbdfio, the ioengine is hardcoded to 'rbd' in the fio_cmd. The uncommenting of this line has no effect and can therefore be reverted.

perezjosibm replied:
Ok, I will remove it

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -31,25 +32,34 @@ def __init__(self, archive_dir, cluster, config):
-#        self.ioengine = config.get('ioengine', 'libaio')
+        self.ioengine = config.get('ioengine', 'libaio')
         self.op_size = config.get('op_size', 4194304)
+        self.set_global_fio_options()
```
I think this call can be moved to `initialize()` and can be called if `self.workloads` is defined. See my suggestion in the comment for the `initialize()` function.

You can also consider renaming this function to `backup_global_fio_options()`

perezjosibm replied:
Ok, makes sense

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -31,25 +32,34 @@ def __init__(self, archive_dir, cluster, config):
         # use_existing_volumes needs to be true to set the pool and rbd names
-        self.use_existing_volumes = config.get('use_existing_volumes', False)
+        self.use_existing_volumes = bool(config.get('use_existing_volumes', False))
+        self.vol_name = config.get('vol_name', 'cbt-librbdfio-`%s`-%d' )
```
`self.vol_name` is unused and therefore can be removed.

perezjosibm replied:
Ah yes, I was experimenting and forgot to remove it, thanks!

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
+        self.op_size = self.global_fio_options['op_size']
+
+
+    def get_workloads(self):
```
This function only logs the workload details. This can be moved to `initialize()`.

perezjosibm replied:
Yeah, I was wondering whether these methods -- get_workloads() -- might need to grow, but probably wont, so I'll remove them. The logging was only to verify that it was being created correctly, so might no longer needed. Or simply as in a debug log.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
+                        pprint.pformat(self.workloads).replace("\n", "\n    "))
+
+
+    def get_precondition(self):
```
This function only logs the precondition details. This can be moved to `initialize()`.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
-        monitoring.start("%s/idle_monitoring" % self.run_dir)
-        time.sleep(60)
-        monitoring.stop()
+        if not self.skip_monitoring:
```
The actual sleep time is different from what is logged. I'd suggest using a new config variable called `idle_monitor_sleep` and use that here. This removes the need for `self.skip_monitoring` boolean.

perezjosibm replied:
I think the flag skip_monitoring is still useful for test where we don't want/need to monitor, for example when pre-filling the volumes, etc.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
         if 'recovery_test' in self.cluster.config and self.recov_test_type == 'background':
             self.mkrecovimage()
+
+        self.get_workloads()
+        self.get_precondition()
```
Following on from one of my previous comments, I'd suggest the following:

```suggestion
    def initialize(self):
       ...
       if (self.workloads):
         self.set_global_fio_options()
         logger.info(" %d Workloads:\n....)
         logger.info("Precond:\n ...)
      ...
```

perezjosibm replied:
I've refactored in an equivalent way, only the precondition has now been removed since was not used, probably a feature for a next PR

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
+            ps = []
+            #test=self.workloads[wk]: aggregate/overwrite the global options
+            test = dict(self.global_fio_options, **self.workloads[wk])
+            enable_monitor=True
```
nit: space before and after '='.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
+            enable_monitor=True
+            logger.info('Running rbd fio %s test, mode %s', wk, test['mode'])
+            if 'monitor' in test:
+                enable_monitor=bool(test['monitor'])
```
nit: Space before and after  '='.

perezjosibm replied:
Yeah, I thought I had run the lint, will double check 👍

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -60,41 +70,120 @@ def __init__(self, archive_dir, cluster, config):
+                        p = common.pdsh(settings.getnodes('clients'), fio_cmd)
+                        ps.append(p)
+                    if enable_monitor:
+                        time.sleep(30); # ramp up time before measuring
```
Use the `self.ramp' here for sleep.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -128,27 +217,34 @@ def run(self):
         common.sync_files('%s/*' % self.run_dir, self.out_dir)
         self.analyze(self.out_dir)
 
+
```
nit: Extraneous new line. Maintain single line gap between function definitions.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -198,6 +297,7 @@ def mkfiocmd(self, volnum):
         fio_cmd += ' %s > %s' % (self.names, out_file)
         return fio_cmd
 
+
```
nit: Extraneous new line.

sseshasa commented on benchmark/librbdfio.py:
```diff
@@ -210,6 +310,7 @@ def mkrecovimage(self):
                   self.cluster.mkimage('cbt-librbdfio-recov-%s-%d' % (node,volnum), self.vol_size, self.recov_pool_name, self.data_pool, self.vol_object_size)
         monitoring.stop()
 
+
```
nit: Extraneous new line. Noticed that a new line is introduced before all functions. Please revert it in all places.

perezjosibm replied:
Will do

sseshasa reviewed (changes requested):
Okay, now my first set of comments are added. Will continue reviewing this.

sseshasa reviewed (changes requested):
I am done with the first pass and have provided my comments. 

- I mainly focused on the main parts of the code changes related to benchmark. 
- One important recommendation is to ensure through testing that the original behavior of librbdfio benchmark is preserved

perezjosibm commented on cbt.py:
```diff
@@ -68,20 +73,23 @@ def main(argv):
             benchmarks = benchmarkfactory.get_all(archive_dir, cluster, iteration)
             for b in benchmarks:
-                if b.exists() and not settings.cluster.get('is_teuthology', False):
+                if not b.exists() and not settings.cluster.get('is_teuthology', False):
```
The change was to fix a bug that was occurring on fresh new runs, but I'm happy to reverted. Will run further tests to confirm  👍

perezjosibm commented on settings.py:
```diff
@@ -116,7 +116,7 @@ def getnodes(*nodelists):
 
     str_nodes = ','.join(uniquenodes(nodes))
-    logger.debug("Nodes : %s", str_nodes)
+    #logger.debug("Nodes : %s", str_nodes)
```
Yes, it floods the log since its called several times by several methods

perezjosibm commented on monitoring.py:
```diff
@@ -149,6 +151,57 @@ def _get_default_nodes():
+                    top_cmd = top_template.format(top_dir=top_dir, pid=pid)
+                    runner = common.sh(local_node, top_cmd)
+                    self.top_runners.append(runner)
+        else:
```
Yes, the 'else' path in this case is for remote nodes 😃

perezjosibm commented on tools/fio-parse-jsons.py:
```diff
@@ -0,0 +1,382 @@
+#!/usr/bin/python3
```
Makes sense, yes I'll do that asap

perezjosibm replied:
Hi @sseshasa: I've opened PR https://github.com/ceph/cbt/pull/307 for only the new script, thanks!

sseshasa replied:
Hi @perezjosibm I mean introduce the new script in a new commit as a part of this PR(#306) itself.

perezjosibm replied:
Hi @sseshasa sorry about that, I hope you don't mind reviewing them separately. I have updated this PR with the requested changes to the best of my understanding. I did a `git ci -s -a --amend` to have them all in a single batch. Many thanks.

perezjosibm commented on common.py:
```diff
@@ -114,15 +114,17 @@ def get_localnode(nodes):
     # if more than one node is listed, fallback to pdsh
     nodes_list = expanded_node_list(nodes)
-    if len(nodes_list) > 1:
+    if len(nodes_list) < 1:
```
Its a bug fix since the logic was returning None when there were actual nodes in the list 😄

perezjosibm commented on cbt.py:
```diff
@@ -68,20 +73,23 @@ def main(argv):
                     b.initialize()
                 # Always try to initialize endpoints before running the test
                 b.initialize_endpoints()
+                logger.info("Running benchmark %s == iteration %d =="%(b,iteration))
```
I've changed to use f-strings instead 👍

perezjosibm commented:
Tested the remote path by running CBT from my laptop, connecting to the remote nodes:
![Screenshot 2024-06-05 at 12 51 09](https://github.com/ceph/cbt/assets/23522684/ccfb08ac-1c97-40f6-86e9-fedab5df1255)

perezjosibm commented:
Apologies, I just fixed some subtle bugs I introduced when doing the changes that didn't show up in the preliminary tests, I need to push the fixes again, sorry for the disturbance!

perezjosibm commented:
Hi @sseshasa changes are ready for review please, many thanks!

sseshasa commented on cluster/cluster.py:
```diff
@@ -5,6 +5,7 @@ def __init__(self, config):
         self.mnt_dir = config.get('mnt_dir', "%s/%s" % (base_tmp, 'mnt'))
         self.tmp_dir = "%s/%s" % (base_tmp, config.get('clusterid'))
         self.archive_dir = "%s/%s" % (config.get('archive_dir'), config.get('clusterid'))
+        #self.idle_monitor_sleep = config.get('idle_monitor_sleep',60)
```
Can this commented line be removed?

perezjosibm replied:
Done, I've updated with the minor fixes 👍

sseshasa reviewed (approved):
Just added a couple of new review comments. But this looks good to me.  I would prefer another pair of eyes to have a look at this as well. @markhpc can you please review this?

perezjosibm commented on benchmark/librbdfio.py:
```diff
@@ -1,16 +1,22 @@
-import settings
-import monitoring
+"""
+    linrbdfio.py -- module to support the FIO benchmark exercising RBD.
```
Good spot, thank you

perezjosibm commented:
Hi @Matan-B wonder if you could please review/approve this PR?, Thanks
